File: back-end/schema/stock_transaction.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import List, Optional
from pydantic import BaseModel, Field

from model.stock_transaction import TransactionTypeEnum
from schema.base import EntityReadSchema, EntityWriteSchema


class StockTransactionDetailCreate(EntityWriteSchema):
    product_id: int
    measurement_id: int
    quantity: int
    unit_price:  float = 0
    currency_id: Optional[int] = None
    exchange_rate: float = 1


class StockTransactionCreate(EntityWriteSchema):
    transaction_type: TransactionTypeEnum
    warehouse_id: int
    to_warehouse_id: Optional[int] = None
    # transaction_date: datetime
    reference_number: Optional[str] = None
    reason: Optional[str] = None

    details: List[StockTransactionDetailCreate]


class base_stock_transaction(EntityWriteSchema):
    product_id: int = Field(..., description="The field product is required")
    warehouse_id: int = Field(..., description="The field warehouse is required")
    measurement_id: Optional[int] = None
    to_warehouse_id: Optional[int] = None
    transaction_type: TransactionTypeEnum
    quantity: int = 0
    reason: Optional[str] = None
    reference_number: Optional[str] = None
    transaction_number: Optional[str] = None


class base_stock_transaction_read(EntityReadSchema):
    product_id: int = Field(..., description="The field product is required")
    warehouse_id: int = Field(..., description="The field warehouse is required")
    measurement_id: Optional[int] = None
    transaction_type: TransactionTypeEnum
    quantity: int = 0
    reason: Optional[str] = None
    reference_number: Optional[str] = None
    transaction_number: Optional[str] = None

File: front-end/shared/types/stocktransaction.ts
export type StockTransactionForm = {
    id: number | null
    warehouse_id: number | null
    measurement_id: number | null
    to_warehouse_id: number | null
    transaction_number?: string
    reference_number?: string
    product_id: number | null
    quantity: number
    reason: string
    transaction_type: string | null
}

export interface StockTransactionDetailCreate {
  product_id: number | null
  measurement_id: number | null
  quantity: number
  unit_price: number
  currency_id: number | null
  total_price: number
  exchange_rate: number
}


export interface StockTransactionCreate {
  id?: number
  warehouse_id: number | null
  to_warehouse_id: number | null
  transaction_type: TransactionType
  // transaction_date: string
  reference_number: string
  transaction_number: string
  reason: string

  details: StockTransactionDetailCreate[]
}

export type TransactionType =
  | 'STOCK_IN'
  | 'STOCK_OUT'
  | 'ADJUSTMENT'
  | 'RETURN'
  | 'DAMAGE'
  | 'QUARANTINE'
  | 'INTERNAL_TRANSFER'

export type Errors = {
    warehouse_id: string | null | undefined,
    measurement_id: string | null | undefined
    product_id: string | null | undefined
    quantity: string | null | undefined
    transaction_type: string | null | undefined
}


export type ValidationDetailItem = {
  loc?: Array<string | number>
  msg?: string
}
